jax/interpreters/pxla.py

```diff
@@ -61,144 +61,138 @@
-
-  def __repr__(self):
-    return ("ShardingSpec(shards_per_axis=%s, is_axis_materialized=%s, "
-            "replication_factors=%s)" %
-            (self.shards_per_axis, self.is_axis_materialized,
-             self.replication_factors))
-
+               sharding: Iterable[AvalDimSharding],
+               mesh_mapping: Iterable[MeshDimAssignment]):
+    self.sharding = tuple(sharding)
+    self.mesh_mapping = tuple(mesh_mapping)
+
+  def indices(self, shape: Tuple[int, ...]) -> np.ndarray:
+    """Returns NumPy-style indices corresponding to a sharding spec.
+
+    Args:
+      shape: The shape of the logical array being sharded.
+
+    Returns:
+      An ndarray with a NumPy-style index for each element of the logical mesh.
+      Each element is an int, a slice object with step=1, or a tuple thereof, to
+      be treated as an index into the full logical array.
+    """
+    assert len(shape) == len(self.sharding)
+
+    axis_indices: List[Sequence[Index]] = []
+    chunk_indices_shape = []
+    for dim, sharding in enumerate(self.sharding):
+      axis_size = shape[dim]
+      if sharding is None:
+        axis_indices.append([slice(None)])
+        # NOTE: We don't append unsharded dimensions here, because they do not appear
+        #       in the mesh mapping.
+      elif isinstance(sharding, Unstacked):
+        assert axis_size == sharding.size, f'{axis_size} != {sharding.size}'
+        axis_indices.append(range(axis_size))
+        chunk_indices_shape.append(axis_size)
+      elif isinstance(sharding, Sharded):
+        shard_size, ragged = divmod(axis_size, sharding.shards)
+        assert not ragged
+        axis_indices.append([slice(i * shard_size, (i + 1) * shard_size)
+                             for i in range(sharding.shards)])
+        chunk_indices_shape.append(sharding.shards)
+      else:
+        assert_unreachable(sharding)
+
+    chunk_indices = np.empty([prod(chunk_indices_shape)], dtype=np.object)
+    for i, idxs in enumerate(it.product(*axis_indices)):
+      chunk_indices[i] = idxs
+    chunk_indices = chunk_indices.reshape(chunk_indices_shape)
+
+    # Ensure that each sharded axis is used exactly once in the mesh mapping
+    num_sharded_dim = len(chunk_indices_shape)
+    sharded_dim_perm = [a.axis for a in self.mesh_mapping if isinstance(a, ShardedAxis)]
+    assert (set(sharded_dim_perm) == set(range(num_sharded_dim)) and
+            len(sharded_dim_perm) == num_sharded_dim)
+    # Replicate/reorder the indices according to the mesh mapping
+    replica_sizes = tuple(a.replicas for a in self.mesh_mapping if isinstance(a, Replicated))
+    replica_dim, sharded_dim = it.count(0), it.count(len(replica_sizes))
+    perm = [next(replica_dim) if isinstance(a, Replicated) else next(sharded_dim)
+            for a in self.mesh_mapping]
+    return (np.broadcast_to(chunk_indices, replica_sizes + chunk_indices.shape)
+              .transpose(perm))
```

Comment:
This trick of putting the replica axes in front and then transposing is very clever :) I don't 100% get it but seems plausible and the tests pass!